<?xml version="1.0"?>
<launch>

    <arg name="r_name" default="racecar"/>
    <arg name="vesc_ns" default="/vesc"/>

    <arg name="max_vel" default="0.5"/>
    <arg name="max_acc" default="0.1"/>

    <arg name="use_comm_layer" default="false"/>

    <param name="max_vel" value="$(arg max_vel)"/>
    <param name="max_acc" value="$(arg max_acc)"/>


    <!--The move base node-->
    <include file="$(find racecar_navigation)/launch/racecar_move_base.launch">
      <arg name="r_name" value="$(arg r_name)"/>
      <arg name="vesc_ns" value="$(arg vesc_ns)"/>
      <arg name="use_comm_layer" value="$(arg use_comm_layer)"/>
    </include>

    
    <!-- Determine current and desired future lanes of racecar -->
    <node pkg="racecar_navigation" type="threeLanes_current_future_pos.py" name="threeLanes_current_future_pos" output="screen" ns="/$(arg r_name)">
       <remap from="/vesc/odom" to="$(arg vesc_ns)/odom"/>
    </node>
    

    <!-- Determine the required navigation action -->
    <node pkg="racecar_navigation" type="customLocalPlanner.py" name="customLocalPlanner" output="screen" ns="/$(arg r_name)"/>


    <!-- Determine whether goal has been reached -->
    <node pkg="racecar_navigation" type="isGoalReached.py" name="isGoalReached" output="screen" ns="/$(arg r_name)">
       <remap from="/vesc/odom" to="$(arg vesc_ns)/odom"/>
    </node>
    

    <!-- The move car action client node -->
    <node pkg="racecar_move_car" type="move_car_action_client.py" name="move_car_action_client" output="screen" ns="/$(arg r_name)"/>


    <!-- The move car lane keeping action server node -->
    <node pkg="racecar_move_car" type="laneKeeping_action_server.py" name="laneKeeping_action_server" output="screen" ns="/$(arg r_name)">
      <remap from="/vesc/odom" to="$(arg vesc_ns)/odom"/>
    </node>
    

    <!-- The move car lane keeping velocity action server node -->
    <node pkg="racecar_move_car" type="laneKeeping_vel_action_server.py" name="laneKeeping_vel_action_server" output="screen" ns="/$(arg r_name)">
      <remap from="/vesc/odom" to="$(arg vesc_ns)/odom"/>
    </node>

    <!-- The move car lane changing action server node -->
    <node pkg="racecar_move_car" type="laneChange_action_server.py" name="laneChange_action_server" output="screen" ns="/$(arg r_name)">
      <remap from="/vesc/odom" to="$(arg vesc_ns)/odom"/>
    </node>
    

    <!-- The drive_param to servo node -->
    <node pkg="racecar_control" type="drive_param_to_servo.py" name="drive_param_to_servo" output="screen" ns="/$(arg r_name)">
       <remap from="/vesc/ackermann_cmd_mux/input/teleop" to="$(arg vesc_ns)/ackermann_cmd_mux/input/teleop"/>
       <remap from="/drive_parameters" to="/$(arg r_name)/drive_parameters"/>
    </node>

</launch>
